torch/distributed/_tensor/redistribute.py

```diff
@@ -1,75 +1,132 @@
 # Copyright (c) Meta Platforms, Inc. and affiliates
-from typing import cast, Dict, List, Tuple
+from functools import lru_cache
+from typing import cast, Dict, List, NamedTuple, Tuple
 
 import torch
 import torch.distributed._tensor.api as dtensor
+from torch.distributed._tensor.device_mesh import DeviceMesh
 from torch.distributed._tensor.placement_types import (
     _Partial,
     DTensorSpec,
     Placement,
     Replicate,
     Shard,
 )
-from torch.distributed.device_mesh import DeviceMesh
 
 
-_PlacementItem = Tuple[int, Tuple[Placement, Placement]]
+class _TransformInfo(NamedTuple):
+    mesh_dim: int
+    src_dst_placements: Tuple[Placement, Placement]
+    # logical_shape on this mesh dimension
+    logical_shape: List[int]
 
 
-def _replicate_then_shard(val: _PlacementItem) -> int:
+def _replicate_then_shard(val: _TransformInfo) -> int:
     """
-    Replicate from inner to outer dimension.
-    Shard from outer to inner dimension.
+    Replicate happens from inner to outer dimension.
+    Shard happens from outer to inner dimension.
     """
-    i, (current, target) = val
-    if (target.is_replicate() or target.is_partial()) and current.is_shard():
-        return -i
-    elif (current.is_replicate() or current.is_partial()) and target.is_shard():
-        return i
+    mesh_dim = val.mesh_dim
+    src, dst = val.src_dst_placements
+    if (dst.is_replicate() or dst.is_partial()) and src.is_shard():
+        return -mesh_dim
+    elif (src.is_replicate() or src.is_partial()) and dst.is_shard():
+        return mesh_dim
     else:
         return 0
 
 
-def _decompose_reshard(val: List[_PlacementItem]) -> List[_PlacementItem]:
+@lru_cache(maxsize=None)
+def _gen_transform_infos(
+    src_spec: DTensorSpec,
+    dst_spec: DTensorSpec,
+) -> List[_TransformInfo]:
     """
-    Decompose Si -> Sj into Si -> R -> Sj
-    There's 2 ways a shardings can differ within a mesh dimension:
-      1) sharding on different tensor dimensions, e.g. Shard(0) -> Shard(1)
-      2) different sub-shards of a repeated shard ("mis-aligned sharding")
-          (Shard(0), Shard(0)) -> (Replicate(), Shard(0))
-          Here the Shard(0) -> Shard(0) for mesh dimension 2 is actually
-          a reshard, because in the first case it's a sub-sharding of an already tensor dimension 0,
-          and in the second case, it's the first sharding on tensor dimension 0.
+    Generate the transform infos from the source placements to the target placements, to
+    transform from source to target placement it might have multipl steps, i.e. it might
+    decompose Si -> Sj into Si -> R -> Sj.
+    This would detects if there're mis-aligned shardings between src/dst placements.
+    i.e. (Shard(0), Shard(0)) -> (Replicate(), Shard(0)), in this case Shard(0) -> Shard(0)
```

Comment:
I think we can add an extra module in future to address this reshard algorithm if we haven't got one yet.

Reply:
No I don't think we should add an "extra" module. `redistribute.py` is already the module that handles redistribution, adding an extra module would make the system more complicated and it's not necessary. 

> It cannot generate the optimal resharding for case (Shard(0), Shard(0)) -> (Replicate(), Shard(0))

what do you mean by "allgather along the second mesh dim"? I think the current logic already gives the best and the valid resharding for this case.